eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/LedgerSpec.hs

```diff
@@ -212,9 +212,35 @@ spec = do
         mkBasicTxBody & withdrawalsTxBodyL .~ Withdrawals [(ra, mempty)]
 
   describe "Mempool" $ do
+    let
+      submitFailingMempoolTx cause tx expectedFailures = do
+        globals <- use impGlobalsL
+        nes <- use impNESL
+        slotNo <- use impLastTickG
+        let
+          mempoolEnv = mkMempoolEnv nes slotNo
+          ls = nes ^. nesEsL . esLStateL
+        case applyTx globals mempoolEnv ls tx of
+          Left err ->
+            err `shouldBe` ApplyTxError @era (injectFailure <$> expectedFailures)
+          Right _ -> assertFailure $ "Expected failure due to " <> cause <> ": " <> show tx
+
+    it "Duplicate transactions" $ do
+      inputs <- replicateM 10 $ do
+        addr <- freshKeyAddr_
+        amount <- Coin <$> choose (2_000_000, 8_000_000)
+        sendCoinTo addr amount
+      tx <- submitTx $ mkBasicTx $ mkBasicTxBody & inputsTxBodyL <>~ fromList inputs
+      submitFailingMempoolTx "duplicate transaction" tx $
+        pure . ConwayMempoolFailure $
+          "All inputs are spent. Transaction has probably already been included"
+      input <- elements inputs
+      let tx1 = mkBasicTx $ mkBasicTxBody & inputsTxBodyL <>~ [input]
+      submitFailingMempoolTx "duplicate transaction" tx1 $
+        pure . ConwayMempoolFailure $
+          "All inputs are spent. Transaction has probably already been included"
```

Comment:
So, this is the proof that your logic in the MEMPOOL rule is incorrect.

This test must produce `BadInputsUTxO (Set.singleton inputs)` predicate failure, not a `ConwayMempoolFailure` failure, because only one input is missing, not all of them.

Reply:
I agree that it should do that, but with the current strategy of checking the spend status of the inputs there isn't any way to distinguish between a transaction that is trying to spend _some_ of the inputs of a preceding transaction or _all_ of them, since we look only at the inputs of the current transaction, not the inputs of a preceding one. In fact, I'm not sure, given an output that's no longer in the UTxO, there's any way to find out which transaction spent it.

Reply:
I'll see if I can make it work with fixups. That would reduce the list of predicate failures in the overlapping case.

I was confused by your suggestion of using a single input, because I wasn't thinking about the fixup.

Reply:
I've updated the tests to reflect clearly what we want the two possible outcomes to be. Maybe you can take a final look and check that it matches what you expect.